Bug fixes, test framework improvements, and new test

robertosassu opened the pull request:
Main features of this PR:

- Improve variable naming of tests
- Move kernel tests to a subdirectory
- Move common functions in kernel tests to a separate library
- Add new test for IMA_NONACTION_RULE_FLAGS
- Fix shellcheck warnings

Changelog
v2:
- Remove change in fsverity.test to suppress the shellcheck warning (provided as a separate patch by Mimi)
- Remove -x option from shellcheck target of tests/Makefile.am and tests/kernel/Makefile.am, and add ../function.sh in the latter (suggested by Mimi)
- Improve commit messages for the kernel library patches (suggested by Mimi)

v1:

- Rebase to https://github.com/linux-integrity/ima-evm-utils/commit/7a4a7ec0c7c5b384bf0f2dfa0d8d9eaf9079fbff ("Merge pull request https://github.com/linux-integrity/ima-evm-utils/pull/16 from hu-wendila/next-testing")
- Execute first user space tests and then kernel space tests (suggested by Mimi)
- Include functions_kernel.sh in shellcheck target
- Locally suppress SC2034 shellcheck warning
- Split getfattr command line in multiple lines to avoid checkpatch error
- Store command exit status in the result variable in get_private_key(), to avoid shellcheck warning
- Put keyring variable between quotes in load_public_key()
- Add verbose messages in non_action_rule_flags.test

robertosassu committed:
Add g_ prefix to global variables in kernel tests

Fix the global variables that don't meet the g_ prefix convention.

Signed-off-by: Roberto Sassu <roberto.sassu@huawei.com>

robertosassu committed:
Move kernel-related tests to the test/kernel subdirectory

Separate user space tests from kernel tests, so that only the former group
of tests can be executed, if the intent is only to test ima-evm-utils.

Minimize the changes by placing the kernel tests in tests/kernel/, a
subdirectory of where the other tests are.

Update the relative path in the kernel tests, to reflect the fact that they
were moved one level down.

One side effect of this change is that gen-keys.sh is executed both in
tests/ and tests/kernel/, since it is required respectively by
sign_verify.test and fsverity.test. To make sure that keys are generated in
the directory where the script was invoked, remove the changing directory
command.

Finally, avoid to print the logs twice in build.sh.

Signed-off-by: Roberto Sassu <roberto.sassu@huawei.com>

robertosassu committed:
Add get_private_key() and load_public_key() to functions_kernel.sh

Move get_private_key() and load_public_key() into the new library,
functions_kernel.sh.

Add the path of the temporary file that will hold a copy of the private key
as a new parameter of get_private_key(), and the path of the public key and
the target kernel keyring as new parameters of load_public_key().

Signed-off-by: Roberto Sassu <roberto.sassu@huawei.com>

robertosassu committed:
Add non_action_rule_flags.test

Add test for commit 57a0ef02fefa ("ima: Reset IMA_NONACTION_RULE_FLAGS
after post_setattr") in the kernel.

Signed-off-by: Roberto Sassu <roberto.sassu@huawei.com>

robertosassu committed:
Suppress shellcheck warnings in tests and tests/kernel

Locally suppress the unused variable warnings in evm_hmac.test and
mmap_check.test.

Also add ../function.sh in tests/kernel/Makefile.am to avoid the SC1091
warnings.

Suggested-by: Mimi Zohar <zohar@linux.ibm.com>
Signed-off-by: Roberto Sassu <roberto.sassu@huawei.com>

robertosassu committed:
Ignore EXECUTE_PERMISSIONS,UNKNOWN_COMMIT_ID,BAD_SIGN_OFF for checkpatch.pl

Ignore EXECUTE_PERMISSIONS because test scripts need to be executable,
UNKNOWN_COMMIT_ID because we are not referring in the commit messages to
kernel commit IDs, and BAD_SIGN_OFF because otherwise there is
the 'Unrecognized email address' error in ci.yml.

Signed-off-by: Roberto Sassu <roberto.sassu@huawei.com>

mimizohar commented:
As much as possible we don't want to turn off different checks - checkpatch.pl or shellcheck. 

- checkpatch.pl: Even with "BAD_SIGN_OFF", it still seems to be detecting a missing or a signature that doesn't match the author, so I'm fine with the change.

- "Suppress shellcheck warnings in tests and tests/kernel" 

A number of warnings are fixed in the next-integrity.fix-warnings branch.  Instead of adding  "shellcheck -x", adding ../functions.sh to the list resolves the shellcheck SC1091 issue.   This patch could be dropped.

mimizohar commented:
Both  "Move get_xattr() and check_load_ima_rule() to functions_kernel.sh" and "Add get_private_key() and load_public_key() to functions_kernel.sh" patches are not just moving code, but define/add new function arguments.   Please update the patch descriptions accordingly.   Nice changes!

robertosassu committed:
Move get_xattr() and check_load_ima_rule() to functions_kernel.sh

To avoid repeating the same code in individual tests, move the common
functions get_xattr() and check_load_ima_rule() into the new library,
function_kernel.sh.

Add the key path and mount point as additional parameters to
check_load_ima_rule(), so that the function can use the key for signing the
policy, and for creating the temporary file in the right place.

Finally, add function_kernel.sh to the shellcheck target.

Signed-off-by: Roberto Sassu <roberto.sassu@huawei.com>
